origins_media/origins_media.module

```diff
@@ -183,3 +196,57 @@ function origins_media_preprocess_fieldset__media_library_widget(array &$variabl
   // Hide the 'Show media item weights' link.
   unset($variables['prefix']['weight_toggle']);
 }
+
+/**
+ * Implements hook_preprocess_media().
+ */
+function origins_media_preprocess_media(array &$variables) {
+  if ($variables['view_mode'] == 'media_library' && $variables['media']->bundle() === 'document') {
+    $media_entity = &$variables['media'];
+    $file_storage = \Drupal::entityTypeManager()->getStorage('file');
+
+    // Get the underlying file associated with this document entity.
+    $file = $file_storage->load($media_entity->field_media_file->target_id);
+    $mimetype = $file->getMimeType();
+
+    // There shouldn't be an image file here, but just in case, screen for it as if
+    // it's genuine it likely has its own thumbnail so we would want to keep that.
+    if (preg_match('/jpeg|jpg|gif|png/', $mimetype)) {
+      return;
+    }
+
+    $mimetype_file = \Drupal::config('media.settings')->get('icon_base_uri') . '/' . str_replace('application/', 'application-', $mimetype) . '.png';
+
+    $thumbnail_file = $file_storage->loadByProperties([
+      'uri' => $mimetype_file,
+    ]);
+    $thumbnail_file = reset($thumbnail_file);
+
+    // Swap in the appropriate file reference for the icon representing this MIME type.
+    $variables['content']['thumbnail'][0]['#item']->setValue(['target_id' => $thumbnail_file->id()]);
+  }
+}
+
+/**
+ * Implements hook_preprocess_image_formatter().
+ */
+function origins_media_preprocess_image_formatter(array &$variables) {
+  if (empty($variables['url']) || $variables['image_style'] != 'thumbnail') {
+    return;
+  }
+
+  // Derive the media entity linked to from the URL.
```

Comment:
This bit feels a bit flaky to me, but it seems to hold up (for now).

There's an underlying API for creating plugins that generate thumbnails for each media entity (each as a managed file) which is more relevant for, say, document previews... but a horrible idea for generalised icons per type. Hence why I've adopted to 'fudge it just before render time' rather than set up a formalised backend plugin for each instance.